File: base/src/main/java/me/soushin/higou/base/app/base/BasePresenter.java
package me.soushin.higou.base.app.base;

import com.jess.arms.mvp.IModel;
import com.jess.arms.mvp.IView;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.CancellationException;

import kotlin.Unit;
import kotlin.coroutines.Continuation;
import kotlin.coroutines.CoroutineContext;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import kotlin.sequences.Sequence;
import kotlinx.coroutines.ChildHandle;
import kotlinx.coroutines.ChildJob;
import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.Dispatchers;
import kotlinx.coroutines.DisposableHandle;
import kotlinx.coroutines.Job;
import kotlinx.coroutines.selects.SelectClause0;

/**
 * @auther SouShin
 * @time 2019/12/19 18:47
 */
public class BasePresenter<M extends IModel, V extends IView> extends com.jess.arms.mvp.BasePresenter
        implements CoroutineScope {

    private Job job;


    @NotNull
    @Override
    public CoroutineContext getCoroutineContext() {
        return Dispatchers.getMain().plus(job);
    }

    @Override
    public void onStart() {
        super.onStart();

    }




}

File: base/src/main/java/me/soushin/higou/base/app/base/BasePresenter.kt
package me.soushin.higou.base.app.base

import com.jess.arms.mvp.BasePresenter
import com.jess.arms.mvp.IModel
import com.jess.arms.mvp.IView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlin.coroutines.CoroutineContext

/**
 * @author created by Soushin
 * @time 2019/12/25 14 38
 */
open class BasePresenter<M :IModel,V:IView> :BasePresenter<M,V>, CoroutineScope {

    constructor(model:M,rootView:V):super(model, rootView)

    constructor(rootView: V):super(rootView)

    constructor()

    private lateinit var job: Job

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main+job

    override fun onStart() {
        super.onStart()
        job = Job()
    }

    override fun onDestroy() {
        super.onDestroy()
        // 关闭页面后，结束所有协程任务
        job.cancel()
    }



}